README
NAME
    Business::CyberSource - Perl interface to the CyberSource Simple Order
    SOAP API

VERSION
    version 0.007010

DESCRIPTION
    This library is a Perl interface to the CyberSource Simple Order SOAP
    API built on Moose and XML::Compile::SOAP technologies. This library
    aims to eventually provide a full interface the SOAPI.

    You may wish to read the Official CyberSource Documentation on Credit
    Card Services for the Simpler Order API
    <http://apps.cybersource.com/library/documentation/dev_guides/CC_Svcs_SO
    _API/html/> as it will provide further information on why what some
    things are and the general workflow.

    To get started you will want to read the documentation in
    Business::CyberSource::Client and Business::CyberSource::Request. If you
    find any documentation unclear or outright missing, please file a bug.

    If there are features that are part of CyberSource's API but are not
    documented, or are missing here, please file a bug. I'll be happy to add
    them, but due to the size of the upstream API, I have not had time to
    cover all the features and some are currently undocumented.

ENVIRONMENT
    all environment variables are prefixed with "PERL_BUSINESS_CYBERSOURCE_"

  DEBUG
    causes all requests to be "carp"ed to STDERR

  Testing
   Credentials
   USERNAME
   PASSWORD
    set's the username and password in the client for running tests.

   Direct Currency Conversion
   DCC_CC_YYYY
    sets the test credit card expiration year for both Visa and MasterCard

   DCC_CC_MM
    sets the test credit card expiration month for both Visa and MasterCard

   DCC_MASTERCARD
    A test credit card number provided by your your credit card processor

   DCC_VISA
    A test credit card number provided by your your credit card processor

EXAMPLE
            use 5.010;
            use Carp;
            use Try::Tiny;

            use Business::CyberSource::Client;
            use Business::CyberSource::Request::Authorization;
            use Business::CyberSource::Request::Capture;

            my $client = Business::CyberSource::Client->new({
                    username   => 'Merchant ID',
                    password   => 'API Key',
                    production => 1,
            });

            my $auth_request = try {
                            Business::CyberSource::Request::Authorization->new({
                                    reference_code => '42',
                                    bill_to => {
                                            first_name  => 'Caleb',
                                            last_name   => 'Cushing',
                                            street      => '100 somewhere st',
                                            city        => 'Houston',
                                            state       => 'TX',
                                            postal_code => '77064',
                                            country     => 'US',
                                            email       => 'xenoterracide@gmail.com',
                                    },
                                    purchase_totals => {
                                            currency => 'USD',
                                            total    => 5.00,
                                    },
                                    card => {
                                            account_number => '4111111111111111',
                                            expiration => {
                                                    month => 9,
                                                    year  => 2025,
                                            },
                                    },
                            });
                    }
                    catch {
                            carp $_;
                    };

            my $auth_response = try {
                            $client->run_transaction( $auth_request );
                    }
                    catch {
                            carp $_;

                            if ( $auth_request->has_trace ) {
                                    carp 'REQUEST: '
                                    . $auth_request->trace->request->as_string
                                    . 'RESPONSE: '
                                    . $auth_request->trace->response->as_string
                                    ;
                            }
                    };

            unless( $auth_response->is_accept ) {
                    carp $auth_response->reason_text;
            }
            else {
                    my $capture_request
                            = Business::CyberSource::Request::Capture->new({
                                    reference_code => $auth_response->reference_code,
                                    service => {
                                            request_id => $auth_response->request_id,
                                    },
                                    purchase_totals => {
                                            total    => $auth_response->auth->amount,
                                            currency => $auth_response->purchase_totals->currency,
                                    },
                            });

                    my $capture_response = try {
                            $client->run_transaction( $capture_request );
                    }
                    catch {
                            carp $_;

                            if ( $capture_request->has_trace ) {
                                    carp 'REQUEST: '
                                    . $capture_request->trace->request->as_string
                                    . 'RESPONSE: '
                                    . $capture_request->trace->response->as_string
                                    ;
                            }
                    };

                    if ( $capture_response->is_accept ) {
                            # you probably want to record this
                            say $capture_response->capture->reconciliation_id;
                    }
            }

    This code is not meant to be DRY, but more of a top to bottom example.
    Also note that if you really want to do Authorization and Capture at one
    time use a Sale. Most common Reasons for Exceptions would be bad input
    into the request object (which validates things) or CyberSource just
    randomly throwing an ERROR, in which case you can usually just retry
    later. You don't have to print the response on error during development,
    you can easily just use the DEBUG Environment variable

ACKNOWLEDGMENTS
    *   Mark Overmeer

        for the help with getting XML::Compile::SOAP::WSS working.

SEE ALSO
    *   Checkout::CyberSource::SOAP

    *   Business::OnlinePayment::CyberSource

BUGS
    Please report any bugs or feature requests on the bugtracker website
    https://github.com/hostgator/business-cybersource/issues or by email to
    development@hostgator.com.

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Caleb Cushing <xenoterracide@gmail.com>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2012 by HostGator.com
    <http://hostgator.com>.

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
Revision history for {{$dist->name}}

{{$NEXT}}
  - Show correct home repository on Metacpan
  - Show the up-to-date repo, now maintained by Hostgator
  - Added ShipTo and associated tests
  - Introduced new ShippingMethod type as enum
  - Added invoiceNumber to RequestPart Item
  - Added ShipFrom RequestPart
  - Added ShipFrom to Request base class
  - Added ShipFrom to unit and author tests
  - Added level 2 / level 3 data fields in purchase totals
  - Added test coverage for level II / level III fields in purchase totals
  - Added OtherTax RequestPart
  - Added OtherTax to Request base class
  - Added OtherTax helper service to breadboard
  - Added tests for OtherTax
  - Added InvoiceHeader RequestPart
  - Added InvoiceHeader helper service to breadboard
  - Added test for InvoiceHeader
  - Modified capture test for InvoiceHeader
  - Added InvoiceHeader credit tests
  - Added InvoiceHeader to Authorization test

0.010007  2016-07-01
  - Added support for sending CommerceIndicator in Auth requests
  - Fixed author tests that were failing based on different CyberSource
    sandbox behaviour rather than due to bugs in the library

0.010006  2015-02-19
	- Fixed filename for production CyberSource Transaction WSDL

0.010005  2014-11-19
	- require Type::Utils version 0.040 for failing cpan test

0.010004  2014-11-18
	- remove  Moose::Util::TypeConstraints from ::Client GH #2

0.010003  2014-06-19
	- use strict mode for ::ProjectDistDir
		- fixes isolated bugs in certain deployments, if you haven't had
		  problems with ::Client locating the wsdl/xsd this hasn't impacted you
	[DEPRECATIONS]
	- added deprecation warnings around alias inializers
	  - these will be removed as if they were put in 0.01000

0.010002  2014-06-18
	- fix Type::Tiny code to use Invocant

0.010001  2014-06-18
	- fix tests for perl 5.20 to not trip on Types::Tiny bug
	- fix SOAPFault documentation

0.010000  2014-06-17
	[NEWS]
	- Converted to Moose::Exception
	  - IMPORTANT read the removal section
	  - NEVER ever check for ->isa('Moose::Exception') because that's an
	    implementation detail and the superclass may change in the future,
	    check final classes or roles as documented instead (MST)
	  - read the updated example in Business::CyberSource
	- Added support for American Express Electronic Verification information
	[REMOVAL]
	- Exception::Base - Some exception objects were renamed
	  - CHECK YOUR EXCEPTION HANDLING, it's probably wrong now, sorry. This
	    shouldn't change again, having external classes means supporting new
	    classes will be easier, existing namespaces were awkward for external
	    classes
	- MooseX::StrictConstructor
	- Path::Class
	[DEPRECATIONS]
	- several undocumented aliases
	- trace, Moose::Exception has trace and this conflicted, renamed
	  http_trace

0.009002  2013-11-29
	- update encoding
	- use Test::Requires for NetAddr::IP tests

0.009001  2013-11-29
	- remove dependency on Class::Load, this is being removed from Moose, use
	  Module::Runtime instead
	- reorganize tests
	- fix documentation

0.009000  2013-08-05
	[NEWS]
	- fix bug with different processor where exceptions were thrown on
	  response
		- remove MooseX::StrictConstructor ( may be readded later )
	[DEPRECATIONS]
	- username, password, production, run_transaction have all been migrated
	  to the MooseY::RemoteHelper::Role::Client user, pass, test, submit
	[REMOVAL]
	- Client->dumper_package, insufficient, hardcode until suitable replacment
	  written
	- previous deprecations
		- billing_info ( use bill_to )
		- accepted is_accepted ( use is_accept )
	[NOTES]
	- due to changes in processor live tests are failing for me, this release
	  fixes the exceptions but I haven't had the time to improve the live test
	  suite to fix false positives.

0.008000  2013-05-21
	[NEWS]
	- API's deprecated in 0.007006 and older have been removed
	[DEPRECATIONS]
	- billing_info ( use bill_to )
	- accepted is_accepted ( use is_accept )

0.007011  2013-05-17
	- JCB Cards now identify as discover in Business::CreditCard 0.32
	- update synopsis example to be more novice friendly with try/catch syntax

0.007010  2012-11-26
	- allow the constructor to treat undefs as unset
		- add depedency on MooseX::UndefTolerant

0.007009  2012-11-19
	- Re-up 0.007008 after fixing pause permissions
	- add Package Metadata

0.007008  2012-11-19
	- BillTo state was not checked as required for US and CA

0.007007  2012-11-15
	- metadata corrections

0.007006  2012-11-15
	[NEWS]
	- Tax Service should now be fully supported
	- massive refactoring of Response objects to use MooseX::RemoteHelper
		- Response uses strict constructor
		- please report any exceptions caused by strict constructor that do not
		  have a status code of 102
	- debug now takes an integer and will be more verbose
		- debug 2 shows full SOAP + hashrefs
		- debug 1 just shows the hashrefs
		- debug 0 shows nothing

	[DEPRECATIONS]
	- change ->dcc->dcc_supported to ->dcc->supported

0.007005  2012-11-01
	- port fixes from 0.006014
	- debug now takes an integer and will be more verbose
		- debug 2 shows full SOAP + hashrefs
		- debug 1 just shows the hashrefs
		- debug 0 shows nothing

0.007004  2012-10-18
	- port fixes from 0.006012
	- insure response object has a trace object

0.007003  2012-10-15
	- delegate more deprecated methods and improve notice

0.007002  2012-10-15
	- documentation improvements
	- more backwards compatability cleanup
	[DEPRECATIONS]
	- is_success entirely

0.007001  2012-10-09
	- remove use of 5.14+ code
	- update documentation
	[DEPRECATIONS]
	- do not check roles in CyberSource Response these are implementation
	  details now

0.007000  2012-10-08
	[NEWS]
	- Tax Service should now be fully supported
	- massive refactoring of Response objects to use MooseX::RemoteHelper
	  improvments
	[DEPRECATIONS]
	- response object is now composed of sub objects
	- response factory api changed

0.006014  2012-11-01

Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;

use File::ShareDir::Install;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl interface to the CyberSource Simple Order SOAP API",
  "AUTHOR" => "Caleb Cushing <xenoterracide\@gmail.com>",
  "BUILD_REQUIRES" => {
    "DateTime" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Test::Fatal" => 0,
    "Test::Moose" => 0,
    "Test::More" => "0.88",
    "Test::Requires" => 0,
    "Test::Requires::Env" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "File::ShareDir::Install" => "0.03"
  },
  "DISTNAME" => "Business-CyberSource",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "Business::CyberSource",
  "PREREQ_PM" => {
    "Class::Load" => "0.20",
    "Config" => 0,
    "DateTime::Format::W3CDTF" => "0.06",
    "Exception::Base" => 0,
    "File::ShareDir::ProjectDistDir" => 0,
    "LWP::Protocol::https" => 0,
    "Module::Load" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::ABC" => "0.06",
    "MooseX::AbstractFactory" => 0,
    "MooseX::Aliases" => 0,
    "MooseX::RemoteHelper" => 0,
    "MooseX::RemoteHelper::CompositeSerialization" => 0,
    "MooseX::SetOnce" => "0.200001",
    "MooseX::StrictConstructor" => 0,
    "MooseX::Types" => 0,
    "MooseX::Types::Common::Numeric" => "0.001003",
    "MooseX::Types::Common::String" => "0.001005",
    "MooseX::Types::CreditCard" => "0.002",
    "MooseX::Types::DateTime" => 0,
    "MooseX::Types::DateTime::W3C" => 0,
    "MooseX::Types::Email" => 0,
    "MooseX::Types::Locale::Country" => 0,
    "MooseX::Types::Locale::Currency" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::NetAddr::IP" => 0,
    "MooseX::Types::Path::Class" => 0,
    "MooseX::UndefTolerant" => 0,
    "Try::Tiny" => 0,
    "XML::Compile::SOAP11" => 0,
    "XML::Compile::SOAP::WSS" => "1.04",
    "XML::Compile::Transport::SOAPHTTP" => 0,
    "XML::Compile::WSDL11" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.007010",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}


